src/UnionFind/Internal.hs

```diff
@@ -0,0 +1,328 @@
+--
+-- FIXME: Implement the optimisation for future calls
+findExn ::
+  Ord key =>
+  key ->
+  UnionFind key value ->
+  (UnionFind key value, key, Maybe value)
+findExn key unionFind =
+  case Map.lookup key $ getMap unionFind of
+    Nothing -> error "findExn: no value associated with key"
+    Just (ChildOf key') -> findExn key' unionFind
+    Just (Ancestor maybeValue) -> (unionFind, key, maybeValue)
+
+-- | @lookup key unionFind@ returns the value associated to @key@ in the
+-- @unionFind@ structure, or @Nothing@ if there is no such binding. It also
+-- returns a new union-find structure optimised for future calls.
+lookup ::
+  Ord key =>
+  key ->
+  UnionFind key value ->
+  (UnionFind key value, Maybe value)
+lookup key unionFind =
+  let (unionFind', _, maybeValue) = find key unionFind
+   in (unionFind', maybeValue)
+
+-- | @union merge unionFind key1 key2@ merges together the equivalence classes
+-- of @key1@ and @key2@ in @unionFind@, thus adding the information that @key1@
+-- and @key2@ are equivalent, and returns an updated union-find structure.
+--
+-- - If neither @key1@ nor @key2@ exists in @unionFind@, they are inserted as
+--   equivalent and without value.
+--
+-- - If either @key1@ or @key2@ exists in @unionFind@ but not the other, the
+--   other is added transparently to the existing one's equivalence class.
+--
+-- - If both @key1@ and @key2@ exist in @unionFind@ in the same equivalence
+--   class, nothing changes, except that @unionFind@ might be optimised for
+--   future calls.
+--
+-- - If both @key1@ and @key2@ exist in @unionFind@ in distinct equivalence
+--   classes, the function @merge@ is called to reconcile the two values
+--   associated to the two equivalence classes. @merge@ receives the value
+--   associated to @key1@'s equivalence class as first argument and the value
+--   associated to @key2@'s equivalence class as second argument.
+union ::
+  Ord key =>
+  (value -> value -> value) ->
+  key ->
+  key ->
+  UnionFind key value ->
+  UnionFind key value
+union merge key1 key2 unionFind =
+  runIdentity $
+    unionM
+      (\value1 value2 -> return $ merge value1 value2)
+      key1
+      key2
+      unionFind
+
+-- | @unionM@ is the same as @union@ but for a monadic merge function.
+unionM ::
+  (Ord key, Monad m) =>
+  (value -> value -> m value) ->
+  key ->
+  key ->
+  UnionFind key value ->
+  m (UnionFind key value)
+unionM merge key1 key2 unionFind =
+  let (unionFind1, ancestor1, maybeValue1) = find key1 unionFind
+      (unionFind2, ancestor2, maybeValue2) = find key2 unionFind1
+   in if ancestor1 == ancestor2
+        then return unionFind2
+        else
+          let onUnionFind2sMap update = return $ UnionFind $ update $ getMap unionFind2
+           in case (maybeValue1, maybeValue2) of
+                (Nothing, Nothing) ->
+                  onUnionFind2sMap $
+                    Map.insert key1 (ChildOf key2)
+                      . Map.insert key2 (Ancestor Nothing)
```

Comment:
`Ancestor` can contain a `Nothing` if you `union` two keys that were unknown until now. That would lead you exactly line 166. However, now that you point me to this function, it looks potentially buggy to me. I think it goes to an inconsistent state if we merge two equivalence classes bound to no value together. As of now, I think it uses the information that `maybeValuen == Nothing` to deduce that `keyn` is unknown to the structure, but that is simply not true. I think it can get fixed by replacing all uses of `keyn` by `ancestorn`, but I need to think about it some more. We should also have tests to catch that sort of mistakes.